# Python: API authentication

**Goal**: Access an API and prepare for scraping!

##  Github API authentication

In general, to access an API, we need to go through **authentication**. This can be done in several ways. We will see the example with the **Github API**.

In [1]:
# Example
import requests

In [2]:
# Creation of the dictionary containing the token
headers = {"Authorization":"token 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"}
nm_github_response = requests.get("https://api.github.com/users/Niangmohamed", headers=headers).json()
nm_github_response

{'login': 'Niangmohamed',
 'id': 45575893,
 'node_id': 'MDQ6VXNlcjQ1NTc1ODkz',
 'avatar_url': 'https://avatars.githubusercontent.com/u/45575893?v=4',
 'gravatar_id': '',
 'url': 'https://api.github.com/users/Niangmohamed',
 'html_url': 'https://github.com/Niangmohamed',
 'followers_url': 'https://api.github.com/users/Niangmohamed/followers',
 'following_url': 'https://api.github.com/users/Niangmohamed/following{/other_user}',
 'gists_url': 'https://api.github.com/users/Niangmohamed/gists{/gist_id}',
 'starred_url': 'https://api.github.com/users/Niangmohamed/starred{/owner}{/repo}',
 'subscriptions_url': 'https://api.github.com/users/Niangmohamed/subscriptions',
 'organizations_url': 'https://api.github.com/users/Niangmohamed/orgs',
 'repos_url': 'https://api.github.com/users/Niangmohamed/repos',
 'events_url': 'https://api.github.com/users/Niangmohamed/events{/privacy}',
 'received_events_url': 'https://api.github.com/users/Niangmohamed/received_events',
 'type': 'User',
 'site_admin':

## Access or endpoints

Let's play with other examples of users or organizations.

In [3]:
# Example by user
k_github_response = requests.get("https://api.github.com/users/karpathy", headers=headers).json()
k_github_response

{'login': 'karpathy',
 'id': 241138,
 'node_id': 'MDQ6VXNlcjI0MTEzOA==',
 'avatar_url': 'https://avatars.githubusercontent.com/u/241138?v=4',
 'gravatar_id': '',
 'url': 'https://api.github.com/users/karpathy',
 'html_url': 'https://github.com/karpathy',
 'followers_url': 'https://api.github.com/users/karpathy/followers',
 'following_url': 'https://api.github.com/users/karpathy/following{/other_user}',
 'gists_url': 'https://api.github.com/users/karpathy/gists{/gist_id}',
 'starred_url': 'https://api.github.com/users/karpathy/starred{/owner}{/repo}',
 'subscriptions_url': 'https://api.github.com/users/karpathy/subscriptions',
 'organizations_url': 'https://api.github.com/users/karpathy/orgs',
 'repos_url': 'https://api.github.com/users/karpathy/repos',
 'events_url': 'https://api.github.com/users/karpathy/events{/privacy}',
 'received_events_url': 'https://api.github.com/users/karpathy/received_events',
 'type': 'User',
 'site_admin': False,
 'name': 'Andrej',
 'company': None,
 'blog'

In [4]:
# Example by organization
m_github_response = requests.get("https://api.github.com/orgs/microsoft", headers=headers).json()
m_github_response

{'login': 'microsoft',
 'id': 6154722,
 'node_id': 'MDEyOk9yZ2FuaXphdGlvbjYxNTQ3MjI=',
 'url': 'https://api.github.com/orgs/microsoft',
 'repos_url': 'https://api.github.com/orgs/microsoft/repos',
 'events_url': 'https://api.github.com/orgs/microsoft/events',
 'hooks_url': 'https://api.github.com/orgs/microsoft/hooks',
 'issues_url': 'https://api.github.com/orgs/microsoft/issues',
 'members_url': 'https://api.github.com/orgs/microsoft/members{/member}',
 'public_members_url': 'https://api.github.com/orgs/microsoft/public_members{/member}',
 'avatar_url': 'https://avatars.githubusercontent.com/u/6154722?v=4',
 'description': 'Open source projects and samples from Microsoft',
 'name': 'Microsoft',
 'company': None,
 'blog': 'https://opensource.microsoft.com',
 'location': 'Redmond, WA',
 'email': 'opensource@microsoft.com',
 'twitter_username': 'OpenAtMicrosoft',
 'is_verified': True,
 'has_organization_projects': True,
 'has_repository_projects': True,
 'public_repos': 4585,
 'public_gi

### Training

Let's apply a GET request to the following access point: **https://api.github.com/repos/karpathy/minGPT**.

In [5]:
minGPT_repos_response = requests.get("https://api.github.com/repos/karpathy/minGPT", headers=headers).json()
minGPT_repos_response

{'id': 288108533,
 'node_id': 'MDEwOlJlcG9zaXRvcnkyODgxMDg1MzM=',
 'name': 'minGPT',
 'full_name': 'karpathy/minGPT',
 'private': False,
 'owner': {'login': 'karpathy',
  'id': 241138,
  'node_id': 'MDQ6VXNlcjI0MTEzOA==',
  'avatar_url': 'https://avatars.githubusercontent.com/u/241138?v=4',
  'gravatar_id': '',
  'url': 'https://api.github.com/users/karpathy',
  'html_url': 'https://github.com/karpathy',
  'followers_url': 'https://api.github.com/users/karpathy/followers',
  'following_url': 'https://api.github.com/users/karpathy/following{/other_user}',
  'gists_url': 'https://api.github.com/users/karpathy/gists{/gist_id}',
  'starred_url': 'https://api.github.com/users/karpathy/starred{/owner}{/repo}',
  'subscriptions_url': 'https://api.github.com/users/karpathy/subscriptions',
  'organizations_url': 'https://api.github.com/users/karpathy/orgs',
  'repos_url': 'https://api.github.com/users/karpathy/repos',
  'events_url': 'https://api.github.com/users/karpathy/events{/privacy}',
  '

## Pagination

**Pagination** is used to manage queries that return a large number of results. 

Let's determine all the repositories of the **first page starred** on the **Github of Karpathy**.

In [6]:
params = {"per_page" : 50, "page" : 1}
p1_starred_repos_response = requests.get("https://api.github.com/users/karpathy/starred", 
                                         headers=headers, 
                                         params=params).json()
p1_starred_repos_response

[{'id': 182281459,
  'node_id': 'MDEwOlJlcG9zaXRvcnkxODIyODE0NTk=',
  'name': 'DeepV2D',
  'full_name': 'princeton-vl/DeepV2D',
  'private': False,
  'owner': {'login': 'princeton-vl',
   'id': 29677746,
   'node_id': 'MDEyOk9yZ2FuaXphdGlvbjI5Njc3NzQ2',
   'avatar_url': 'https://avatars.githubusercontent.com/u/29677746?v=4',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/princeton-vl',
   'html_url': 'https://github.com/princeton-vl',
   'followers_url': 'https://api.github.com/users/princeton-vl/followers',
   'following_url': 'https://api.github.com/users/princeton-vl/following{/other_user}',
   'gists_url': 'https://api.github.com/users/princeton-vl/gists{/gist_id}',
   'starred_url': 'https://api.github.com/users/princeton-vl/starred{/owner}{/repo}',
   'subscriptions_url': 'https://api.github.com/users/princeton-vl/subscriptions',
   'organizations_url': 'https://api.github.com/users/princeton-vl/orgs',
   'repos_url': 'https://api.github.com/users/princeton-vl/repo

### Training

Let's determine all the repositories of the **second page starred** on the **Github of Karpathy**.

In [7]:
params = {"per_page" : 50, "page" : 2}
p2_starred_repos_response = requests.get("https://api.github.com/users/karpathy/starred", 
                                         headers=headers, 
                                         params=params).json()
p2_starred_repos_response

[{'id': 20346459,
  'node_id': 'MDEwOlJlcG9zaXRvcnkyMDM0NjQ1OQ==',
  'name': 'HackerNewsData',
  'full_name': 'sytelus/HackerNewsData',
  'private': False,
  'owner': {'login': 'sytelus',
   'id': 2096835,
   'node_id': 'MDQ6VXNlcjIwOTY4MzU=',
   'avatar_url': 'https://avatars.githubusercontent.com/u/2096835?v=4',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/sytelus',
   'html_url': 'https://github.com/sytelus',
   'followers_url': 'https://api.github.com/users/sytelus/followers',
   'following_url': 'https://api.github.com/users/sytelus/following{/other_user}',
   'gists_url': 'https://api.github.com/users/sytelus/gists{/gist_id}',
   'starred_url': 'https://api.github.com/users/sytelus/starred{/owner}{/repo}',
   'subscriptions_url': 'https://api.github.com/users/sytelus/subscriptions',
   'organizations_url': 'https://api.github.com/users/sytelus/orgs',
   'repos_url': 'https://api.github.com/users/sytelus/repos',
   'events_url': 'https://api.github.com/users/sytel

## User-level access point

The **user-level access point** refers to the fact that the server only keeps track of the user who has logged in with a well-defined token. So once we authenticate, we don't need to specify the name of the user, we can put user directly and the query will interpret user directly as the user who authenticated.

In [8]:
# Example
user_access_point = requests.get("https://api.github.com/user", headers=headers).json() # user=Niangmohamed
user_access_point

{'login': 'Niangmohamed',
 'id': 45575893,
 'node_id': 'MDQ6VXNlcjQ1NTc1ODkz',
 'avatar_url': 'https://avatars.githubusercontent.com/u/45575893?v=4',
 'gravatar_id': '',
 'url': 'https://api.github.com/users/Niangmohamed',
 'html_url': 'https://github.com/Niangmohamed',
 'followers_url': 'https://api.github.com/users/Niangmohamed/followers',
 'following_url': 'https://api.github.com/users/Niangmohamed/following{/other_user}',
 'gists_url': 'https://api.github.com/users/Niangmohamed/gists{/gist_id}',
 'starred_url': 'https://api.github.com/users/Niangmohamed/starred{/owner}{/repo}',
 'subscriptions_url': 'https://api.github.com/users/Niangmohamed/subscriptions',
 'organizations_url': 'https://api.github.com/users/Niangmohamed/orgs',
 'repos_url': 'https://api.github.com/users/Niangmohamed/repos',
 'events_url': 'https://api.github.com/users/Niangmohamed/events{/privacy}',
 'received_events_url': 'https://api.github.com/users/Niangmohamed/received_events',
 'type': 'User',
 'site_admin':

## POST request

The goal of the **POST** request is to **send informations (create an object)** to the **servers API**. For example, for Github, you can use it to create repositories. Note that for each request, the API is free to accept it or not.

In [9]:
# Example
repos = {"name":"API-Authentication-in-Python"}
add_repos_response = requests.post("https://api.github.com/user/repos", json=repos, headers=headers)

In [10]:
add_repos_response_status = add_repos_response.status_code
add_repos_response_status

201

The code **201** shows that the request worked and created the github *API-Authentication-in-Python* repository.

## PATCH / PUT request

The **PATCH** request is used to change attributes of an object. For example changing the name or description of a repository. And we use the **PUT** request when we want to change the object completely. All these requests depend on the **APIs** used. Some **APIs** can accept only the **PUT** request, other **APIs** can accept a **PUT** request as a **PATCH** request.

In [11]:
# Example
repos_edited = {
    "name":"API-Authentication", 
    "description":"Access and request an API and prepare for scraping!"}
modify_repos_response = requests.patch("https://api.github.com/repos/Niangmohamed/API-Authentication-in-Python", 
                                    json=repos_edited, headers=headers)

In [12]:
modify_repos_response_status = modify_repos_response.status_code
modify_repos_response_status

200

In github, the **PUT** request works the same way as **PATCH**.

## DELETE request

The **DELETE** request allows to delete an object from an API.

In [13]:
delete_repos_response = requests.delete("https://api.github.com/repos/Niangmohamed/API-Authentication-in-Python", 
                                       headers=headers)

In [14]:
delete_repos_response_status = delete_repos_response.status_code
delete_repos_response_status

204